go/perf/suite/suite.go

```diff
@@ -0,0 +1,334 @@
+	Database datas.Database
+
+	// DatabaseSpec is the Noms spec of `Database` (typically a localhost URL).
+	DatabaseSpec string
+
+	tempFiles []*os.File
+	paused    time.Duration
+}
+
+type perfSuiteT interface {
+	Suite() *PerfSuite
+}
+
+type timeInfo struct {
+	elapsed, paused, total time.Duration
+}
+
+type testRun map[string]timeInfo
+
+type nopWriter struct{}
+
+func (r nopWriter) Write(p []byte) (int, error) {
+	return len(p), nil
+}
+
+// Run runs `suiteT` and writes results to dataset `datasetID` in the database given by the `-perf` command line flag.
+func Run(datasetID string, t *testing.T, suiteT perfSuiteT) {
+	if *perfFlag == "" {
+		return
+	}
+
+	assert := assert.New(t)
+
+	// Piggy-back off the go test -v flag.
+	verboseFlag := flag.Lookup("test.v")
+	assert.NotNil(verboseFlag)
+	verbose := verboseFlag.Value.(flag.Getter).Get().(bool)
+
+	suite := suiteT.Suite()
+	suite.T = t
+	if verbose {
+		suite.W = os.Stdout
+	} else {
+		suite.W = nopWriter{}
+	}
+
+	gopath := os.Getenv("GOPATH")
+	assert.True(gopath != "")
+	suite.AtticLabs = path.Join(gopath, "src", "github.com", "attic-labs")
+
+	// This is the database the perf test results are written to.
+	db, err := spec.GetDatabase(*perfFlag)
+	assert.NoError(err)
+
+	// This is the temporary database for tests to use.
+	server := datas.NewRemoteDatabaseServer(chunks.NewMemoryStore(), 0)
+	portChan := make(chan int)
+	server.Ready = func() { portChan <- server.Port() }
+	go server.Run()
+
+	port := <-portChan
+	suite.DatabaseSpec = fmt.Sprintf("http://localhost:%d", port)
+	suite.Database = datas.NewRemoteDatabase(suite.DatabaseSpec, "")
+
+	// List of test runs, each a map of test name => timing info.
+	testRuns := make([]testRun, *perfRepeatFlag)
+
+	defer func() {
+		for _, f := range suite.tempFiles {
+			os.Remove(f.Name())
+		}
+
+		runs := make([]types.Value, *perfRepeatFlag)
+		for i, run := range testRuns {
+			timesSlice := []types.Value{}
+			for name, info := range run {
+				timesSlice = append(timesSlice, types.String(name), types.NewStruct("", types.StructData{
+					"elapsed": types.Number(info.elapsed.Nanoseconds()),
+					"paused":  types.Number(info.paused.Nanoseconds()),
+					"total":   types.Number(info.total.Nanoseconds()),
+				}))
+			}
+			runs[i] = types.NewMap(timesSlice...)
+		}
+
+		record := types.NewStruct("", map[string]types.Value{
+			"environment":     suite.getEnvironment(),
+			"nomsVersion":     types.String(suite.getGitHead(path.Join(suite.AtticLabs, "noms"))),
+			"testdataVersion": types.String(suite.getGitHead(path.Join(suite.AtticLabs, "testdata"))),
+			"runs":            types.NewList(runs...),
+		})
+
+		ds := dataset.NewDataset(db, datasetID)
+		var err error
+		ds, err = ds.CommitValue(record)
+		assert.NoError(err)
+		assert.NoError(db.Close())
+		server.Stop()
+	}()
+
+	for runIdx := 0; runIdx < *perfRepeatFlag; runIdx++ {
+		run := testRun{}
```

Comment:
`run` seems mostly unneeded here, and while this works fine because it's a Go reference type, that fact is obscured by using a type alias (or whatever Go calls them) here. I had to go remind myself that `testRun` is _just_ a Go map to be sure this works OK.

I'd suggest just using `testRuns[runIdx]` in this loop, since you only use it once anyway.